### PR TITLE
File upload

### PR DESCRIPTION
**Pull request by vojta7**
This is my WIP for file uploading #6 

I would like to clean up my error handling and discus some ideas.

- if we can upload files, maybe I should add option to delete files too
- Is it ok to override existing files?
- I think the ability to upload file should be behind CLI flag and not by default

I've tested my code only under linux, but it seems ok.

### CODE REVIEW

**Comment by boastful-squirrel** [comment]
I think there should indeed be an argument in the CLI to allow uploading.

**Comment by svenstaro** [comment]
You should also advertise the uploading feature in the README.

File: src/file_upload.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+use std::{
+    io::Write,
+    path::{Component, PathBuf},
+};
+
+pub fn save_file(
+    field: multipart::Field<dev::Payload>,
+    file_path: PathBuf,
+) -> Box<Future<Item = i64, Error = Error>> {
+    let mut file = match std::fs::File::create(file_path) {
```
Comment (svenstaro) [inline src/file_upload.rs]: Can't you use `?` here?

File: src/file_upload.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                .and_then(|content_disposition| {
+                    content_disposition
+                        .get_filename()
+                        .ok_or(err())
+                        .map(|cd| String::from(cd))
+                });
+            match filename {
+                Ok(f) => {
+                    file_path = file_path.join(f);
+                    // TODO should I allow overriding existing files?
```
Comment (svenstaro) [inline src/file_upload.rs]: Make this configurable via a flag but off by default.

File: src/file_upload.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            mp.map_err(error::ErrorInternalServerError)
+                .map(move |item| handle_multipart(item, file_path.clone()))
+                .flatten(),
+        ),
+    }
+}
+
+/// Handle incoming request to upload file.
+/// Target file path is expected as path parameter in URI and is interpreted as relative from
+/// server root directory. Any path whitch will go outside of this directory is considered
```
Comment (svenstaro) [inline src/file_upload.rs]: "whitch" typo

File: src/file_upload.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    http::header::{ContentDisposition, LOCATION},
+    multipart, Error, FutureResponse, HttpMessage, HttpRequest, HttpResponse,
+};
+use futures::{future, Future, Stream};
+use std::{
+    io::Write,
+    path::{Component, PathBuf},
+};
+
+pub fn save_file(
```
Comment (svenstaro) [inline src/file_upload.rs]: Document this public function.

File: src/file_upload.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                    .write_all(bytes.as_ref())
+                    .map(|_| acc + bytes.len() as i64)
+                    .map_err(|e| error::MultipartError::Payload(error::PayloadError::Io(e)));
+                future::result(rt)
+            })
+            .map_err(|e| error::ErrorInternalServerError(e)),
+    )
+}
+
+pub fn handle_multipart(
```
Comment (svenstaro) [inline src/file_upload.rs]: Document this public function.

File: src/file_upload.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    }
+}
+
+/// Handle incoming request to upload file.
+/// Target file path is expected as path parameter in URI and is interpreted as relative from
+/// server root directory. Any path whitch will go outside of this directory is considered
+/// invalid.
+/// This method returns future.
+pub fn upload_file(req: &HttpRequest<crate::MiniserveConfig>) -> FutureResponse<HttpResponse> {
+    if !req.query().contains_key("path") {
```
Comment (svenstaro) [inline src/file_upload.rs]: Use proper typed param parsing here as we've done in other parts of this crate.

File: src/file_upload.rs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        return Box::new(future::ok(
+            HttpResponse::BadRequest().body("Unspecified parameter path"),
+        ));
+    }
+    // server root path should be valid so we can unwrap it
+    let app_root_dir = req.state().path.clone().canonicalize().unwrap();
+
+    let path_str = req.query()["path"].clone();
```
Comment (svenstaro) [inline src/file_upload.rs]: Again, use proper typed query params.

File: src/main.rs [hunk 8]
```diff
@@ -195,6 +197,11 @@ fn configure_app(app: App<MiniserveConfig>) -> App<MiniserveConfig> {
     let random_route = app.state().random_route.clone().unwrap_or_default();
     let full_route = format!("/{}", random_route);
 
+    //allow file upload
```
Comment (svenstaro) [inline src/main.rs]: Add space before text.

**Comment by boastful-squirrel** [comment]
Shouldn't we disable the upload feature by default, so the user has to explicitly allow it? I fear a future issue "someone uploaded tons of crap on my computer because upload is enabled by default and I didn't need it at all".

**Comment by svenstaro** [comment]
Yes, that is the right way.

**Comment by svenstaro** [comment]
It should also check whether it can even write to the current dir.

File: src/file_upload.rs (hunk 1 above)
Comment (vojta7) [reply src/file_upload.rs]: As far as I know using `?` in function returning future is not possible.

**Comment by vojta7** [comment]
Right know file uploading is not working when used with random path, I am working on fixing it.

I still need to figure out how to properly communicate internal errors, such as conflict with existing file, to end user.

I will advertise file uploading in README, after i fix all these issues.

**Comment by svenstaro** [comment]
We use the failure crate to make error handling a little easier. Check it
out. You can probably use it to ease your error handling.

File: src/file_upload.rs (hunk 1 above)
Comment (svenstaro) [reply src/file_upload.rs]: True. I just checked and sadly it's not possible.

File: src/main.rs [hunk 9]
```diff
@@ -173,11 +181,12 @@ fn main() {
 }
 
 /// Configures the Actix application
-fn configure_app(app: App<MiniserveConfig>) -> App<MiniserveConfig> {
+fn configure_app(mut app: App<MiniserveConfig>) -> App<MiniserveConfig> {
```
Comment (svenstaro) [inline src/main.rs]: Do you really need this mut?

Comment (vojta7) [reply src/main.rs]: Probably not, I changed it and I think it is better now.

**Comment by vojta7** [comment]
I think that I fixed most of errors and it is starting to work quite nice.

There is still little problem with longer uploads, that the server will fail to redirect you back to original page, but uploading finishes successfully. I am not really sure how to solve this right now, but I will look into it.

Error such as trying to override file are now propagated to user.

Visual aspect is lacking, but since @boastful-squirrel seems to be working on visual rework and my designer skill are not good =D, it is probably better to leave it for now.

I would be glad if you could test it (especially on windows and OS X) and tell me, how you like it so far.

File: src/errors.rs [hunk 10]
```diff
@@ -1,6 +1,26 @@
...
+    #[fail(display = "Failed to parse incoming request")]
+    ParseError,
+    /// This error will occur when we fail to precess multipart request
+    #[fail(display = "Failed to process multipart request")]
+    MultipartError(actix_web::error::MultipartError),
+    /// This error may occur when trying to write incoming file to disk
+    #[fail(display = "Failed to create or write to file")]
+    IOError(std::io::Error),
+    /// This error will occur when we he have insuffictent permissions to create new file
+    #[fail(display = "Insuffitient permissions to create file")]
```
Comment (svenstaro) [inline src/errors.rs]: Fix spelling: Insufficient.

File: src/errors.rs [hunk 11]
```diff
@@ -1,6 +1,26 @@
 use failure::{Backtrace, Context, Fail};
 use std::fmt::{self, Debug, Display};
 
+/// Kinds of errors which might happen during file upload
+#[derive(Debug, Fail)]
+pub enum FileUploadErrorKind {
+    /// This error will occur when file overriding is off and file with same name already exists
```
Comment (svenstaro) [inline src/errors.rs]: ` and a file with `

File: src/errors.rs [hunk 12]
```diff
@@ -1,6 +1,26 @@
...
+    #[fail(display = "File with this name already exists")]
+    FileExist,
+    /// This error will occur when server will fail to preccess http header during file upload
```
Comment (svenstaro) [inline src/errors.rs]: `This error will occur when the server fails to process the HTTP header during file upload`

File: src/errors.rs [hunk 13]
```diff
@@ -1,6 +1,26 @@
...
+    #[fail(display = "Failed to parse incoming request")]
+    ParseError,
+    /// This error will occur when we fail to precess multipart request
```
Comment (svenstaro) [inline src/errors.rs]: `This error will occur when we fail to process the multipart request`

File: src/errors.rs [hunk 14]
```diff
@@ -1,6 +1,26 @@
...
+    #[fail(display = "Failed to process multipart request")]
+    MultipartError(actix_web::error::MultipartError),
+    /// This error may occur when trying to write incoming file to disk
```
Comment (svenstaro) [inline src/errors.rs]: `This error may occur when trying to write the incoming file to disk`

**Comment by svenstaro** [comment]
Also, it doesn't build right now :P

File: src/file_upload.rs [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+                query.path.clone()
+            }
+        }
+        Err(_) => {
+            return Box::new(future::ok(
+                HttpResponse::BadRequest().body("Unspecified parameter path"),
+            ))
+        }
+    };
+    // this is really ugly I will try to think about something smarter
```
Comment (svenstaro) [inline src/file_upload.rs]: I agree. Maybe there is a static way to get this header?

**Comment by svenstaro** [comment]
@boastful-squirrel Could you also take a look at this PR?

I kinda regret not having integration tests now! :P

**Comment by vojta7** [comment]
> Also, it doesn't build right now :P

Do you have any idea why it building ok on my PC in Travis CI?

**Comment by svenstaro** [comment]
Maybe your version of rust nightly is somewhat dated? For the error, you can check https://travis-ci.org/svenstaro/miniserve/jobs/513105792

**Comment by vojta7** [comment]
There was missing closing bracket in my code, by I don't understand how is it possible. It was succesfully building before I updated my rustc and then it showed me same error as Travis CI.

File: src/file_upload.rs (hunk 7 above)
Comment (vojta7) [reply src/file_upload.rs]: It's fixed now.

**Comment by boastful-squirrel** [comment]
I'll check asap.

**Comment by svenstaro** [comment]
Please also run clippy on your PR. I just did and I find lots of suggestions that would simplify the code.

File: src/renderer.rs [hunk 16]
```diff
@@ -14,12 +14,22 @@ pub fn page(
...
+    current_dir: &str,
 ) -> Markup {
     html! {
         (page_header(page_title))
         body {
             span #top { }
-            h1.title { (page_title) }
+            h1 { (page_title) }
+            @if file_upload {
+            form action={(upload_route) "?path=" (current_dir)} method="POST" enctype="multipart/form-data" {
```
Comment (svenstaro) [inline src/renderer.rs]: Please indent this section properly.

File: src/args.rs [hunk 17]
```diff
@@ -47,6 +47,14 @@ struct CLIArgs {
     /// Do not follow symbolic links
     #[structopt(short = "P", long = "no-symlinks")]
     no_symlinks: bool,
+
+    /// Enable file uploading
+    #[structopt(short = "u", long = "upload-files")]
+    file_upload: bool,
+
+    /// Enable overriding existing files during file upload
+    #[structopt(short = "o", long = "owerride-files")]
```
Comment (svenstaro) [inline src/args.rs]: It's `overwrite-files`.

File: src/args.rs [hunk 18]
```diff
@@ -100,5 +108,7 @@ pub fn parse_args() -> crate::MiniserveConfig {
         path_explicitly_chosen,
         no_symlinks: args.no_symlinks,
         random_route,
+        override_files: args.override_files,
```
Comment (svenstaro) [inline src/args.rs]: Please also call it `overwrite_files`.

**Comment by vojta7** [comment]
Thank you for pointing out these typos. I am do them quite often and I am really blind when it comes to finding them.

**Comment by svenstaro** [comment]
I would also like it if you were able to just dragndrop files onto the page for the current dir. I think that would make it a bit neater still. Can you add that?

**Comment by svenstaro** [comment]
Also any chance to fix https://github.com/svenstaro/miniserve/pull/52/files#r270477369?

File: src/file_upload.rs [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+    let return_path: String = req.headers()[header::REFERER]
+        .clone()
+        .to_str()
+        .unwrap_or("/")
+        .to_owned();
+    let r_p2 = return_path.clone();
+
+    // if target path is under app root directory save file
```
Comment (svenstaro) [inline src/file_upload.rs]: `    // If the target path is under the app root directory, save the file.`

**Comment by vojta7** [comment]
> Also any chance to fix https://github.com/svenstaro/miniserve/pull/52/files#r270477369?

I still didn't found better solution.

I added drag and drop with a little bit of js, but simply submitting the file works just fine even without js. Right now it is not super clear where to drop file (main table with all files and dirs) and I am not sure how to improve it.

**Comment by svenstaro** [comment]
Ok don't worry about that bit, we'll figure it out later. Can you resolve the conflicts/rebase so I can merge?

**Comment by svenstaro** [comment]
One thing I noticed is that currently, you're not able to drag the file anywhere, you can only drag it into the file listing area. I think dragging it anywhere on the page should work.

Also, there should be some kind of overlay indication that will let you know that "yes, dropping that file now will result in an upload". Currently you don't really know if dragndrop will result in an upload.

Can you also give the user a written hint somewhere that either upload form or dragndrop works?

**Comment by vojta7** [comment]
I am not exactly sure what I did. I did not mean to close this, but played with git a little bit and it resulted into this.

I changed drop zone to work on whole page and added a visual indication. Still not sure how to indicate successful upload.

**Comment by vojta7** [comment]
I tried to follow these articles
[reopen pull request](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c)
[reopen pull request](https://github.com/isaacs/github/issues/361)
but none of it works for me, I just can't reopen this pull request. Do you have any idea what to do now?

**Comment by svenstaro** [comment]
I can't reopen it. Button is grayed out.

**Comment by boastful-squirrel** [comment]
I guess you can create a new branch from your master, merge `file_upload` into your new branch, push, and that should ask you to create a new PR, right ?

**Pull request closed without merging**